File: backend/app/services/classifier.py
from __future__ import annotations

from app.category_presets import CATEGORY_PRESETS, UNCATEGORIZED_SLUG


# Default keyword definitions covering every built-in preset. Callers that know
# the user's active category selection should pass their own defs instead.
DEFAULT_CATEGORY_DEFS: list[dict] = [
    {"name": preset["slug"], "keywords": tuple(preset["keywords"])}
    for preset in CATEGORY_PRESETS
]


def classify_title(title: str, category_defs: list[dict] | None = None) -> str:
    """Classify from the saved-post title only.

    Body text and OCR are intentionally ignored because they are unavailable or
    unreliable for many video posts. Ties are broken by ``category_defs`` order.
    """
    defs = category_defs if category_defs is not None else DEFAULT_CATEGORY_DEFS
    haystack = (title or "").strip().lower()
    best_category = UNCATEGORIZED_SLUG
    best_score = 0

    for definition in defs:
        keywords = definition.get("keywords") or ()
        score = sum(1 for keyword in keywords if keyword and keyword.lower() in haystack)
        if score > best_score:
            best_score = score
            best_category = definition["name"]

    return best_category


def classify_text(
    title: str,
    raw_text: str | None = None,
    ocr_text: str | None = None,
    category_defs: list[dict] | None = None,
) -> tuple[str, str | None]:
    """Backward-compatible wrapper returning ``(category, sub_category)``."""
    del raw_text, ocr_text
    return classify_title(title, category_defs), None

File: extension/src/lib/classifier.js
// Deterministic keyword classifier. Ported from backend/app/services/classifier.py.
import { CATEGORY_PRESETS, UNCATEGORIZED_SLUG } from "./categories.js";

// [{ name, keywords }] covering every built-in preset.
export const DEFAULT_CATEGORY_DEFS = CATEGORY_PRESETS.map((p) => ({
  name: p.slug,
  keywords: p.keywords,
}));

// Classify from the title only. Ties are broken by categoryDefs order.
export function classifyTitle(title, categoryDefs) {
  const defs = categoryDefs || DEFAULT_CATEGORY_DEFS;
  const haystack = String(title || "").trim().toLowerCase();
  let best = UNCATEGORIZED_SLUG;
  let bestScore = 0;
  for (const def of defs) {
    const keywords = def.keywords || [];
    let score = 0;
    for (const kw of keywords) {
      if (kw && haystack.includes(String(kw).toLowerCase())) score += 1;
    }
    if (score > bestScore) {
      bestScore = score;
      best = def.name;
    }
  }
  return best;
}

// Build classification defs from a config object { selected_category_slugs, custom_categories }.
export function classificationDefs(config) {
  const selected = new Set(config?.selected_category_slugs || []);
  const defs = [];
  for (const preset of CATEGORY_PRESETS) {
    if (selected.has(preset.slug)) defs.push({ name: preset.slug, keywords: preset.keywords });
  }
  for (const entry of config?.custom_categories || []) {
    const name = String(entry.name || entry.slug || "").trim();
    const keywords = (entry.keywords || []).map((k) => String(k).trim()).filter(Boolean);
    if (name && keywords.length) defs.push({ name, keywords });
  }
  return defs;
}

// Active categories (enabled presets + custom + Uncategorized) for dropdowns/filters.
export function activeCategories(config) {
  const selected = new Set(config?.selected_category_slugs || []);
  const out = [];
  for (const preset of CATEGORY_PRESETS) {
    if (selected.has(preset.slug)) out.push({ ...preset, is_custom: false });
  }
  const presetSlugs = new Set(CATEGORY_PRESETS.map((p) => p.slug));
  for (const entry of config?.custom_categories || []) {
    const name = String(entry.name || entry.slug || "").trim();
    if (name && !presetSlugs.has(name)) {
      out.push({
        slug: name,
        label_zh: name,
        label_en: name,
        keywords: (entry.keywords || []).map((k) => String(k).trim()).filter(Boolean),
        is_custom: true,
      });
    }
  }
  out.push({ slug: "Uncategorized", label_zh: "未分类", label_en: "Uncategorized", keywords: [], is_custom: false });
  return out;
}
